core/pkg/monitor/system.go

```diff
@@ -205,102 +269,93 @@ func (s *System) Sample() (*spb.StatsRecord, error) {
 	return marshal(metrics, timestamppb.Now()), errors.Join(errs...)
 }
 
-// collectNetworkMetrics gathers network traffic statistics.
-func (s *System) collectNetworkMetrics(metrics map[string]any) error {
-	netIOCounters, err := net.IOCounters(false)
-	if err != nil {
+// collectProcessTreeMetrics gathers RSS, CPU%, and thread count for a process and its descendants.
+func (s *System) collectProcessTreeMetrics(
+	root *process.Process,
+	virtualMem *mem.VirtualMemoryStat,
+	metrics map[string]any,
+) error {
+	// Safeguard to prevent processAndDescendants from taking too long.
+	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+	defer cancel()
+	procs, err := s.processAndDescendants(ctx, root.Pid)
+
+	if err != nil && !errors.Is(err, context.DeadlineExceeded) {
 		return err
 	}
 
-	if len(netIOCounters) > 0 {
-		metrics["network.sent"] = float64(int(netIOCounters[0].BytesSent) - s.networkBytesSentInit)
-		metrics["network.recv"] = float64(int(netIOCounters[0].BytesRecv) - s.networkBytesRecvInit)
+	if len(procs) == 0 {
+		return fmt.Errorf("system: empty process tree")
 	}
 
-	return nil
-}
-
-// collectSystemMemoryMetrics gathers system-wide memory statistics.
-func (s *System) collectSystemMemoryMetrics(metrics map[string]any) (*mem.VirtualMemoryStat, error) {
-	virtualMem, err := mem.VirtualMemory()
-	if err != nil {
-		return nil, err
-	}
+	var (
+		totalRSS     uint64
+		totalCPU     float64
+		totalThreads int32
+	)
 
-	// Total system memory usage in percent
-	metrics["memory_percent"] = virtualMem.UsedPercent
-	// Total system memory available in MB
-	metrics["proc.memory.availableMB"] = float64(virtualMem.Available) / 1024 / 1024
+	for _, p := range procs {
+		// Memory
+		if mi, err := p.MemoryInfo(); err == nil {
```

Comment:
could you leave comment that use the err == nil, as  a convention we should check not nil